File: lego-osek-master/crv/systems/lego-osek/properties/deadzone_symmetry/DeadzoneSymmetryMon.hs
module Main where

import Language.Copilot
import Copilot.Compile.C99
import Data.Int (Int32)

import Prelude hiding ((&&), (||), not, (==), (/=), (<=), (>=))

fr_steering_angle :: Stream Int32
fr_steering_angle = extern "fr_steering_angle" Nothing

fr_motor_left :: Stream Int32
fr_motor_left = extern "fr_motor_left" Nothing

fr_motor_right :: Stream Int32
fr_motor_right = extern "fr_motor_right" Nothing

spec :: Spec
spec = do
  let inDeadzone = (fr_steering_angle <= 2) && (fr_steering_angle >= (-2))
  let motorsEqual = fr_motor_left == fr_motor_right
  let p = (not inDeadzone) || motorsEqual
  trigger "deadzone_symmetry_violation" (not p) []

main :: IO ()
main = reify spec >>= compile "deadzone_symmetry_mon"

File: lego-osek-master/crv/systems/lego-osek/properties/deadzone_symmetry/deadzone_symmetry_mon.c
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#include "deadzone_symmetry_mon_types.h"
#include "deadzone_symmetry_mon.h"

static int32_t fr_steering_angle_cpy;
static int32_t fr_motor_left_cpy;
static int32_t fr_motor_right_cpy;

static bool deadzone_symmetry_violation_0_guard(void) {
  return !((!(((fr_steering_angle_cpy) <= ((int32_t)(2))) && ((fr_steering_angle_cpy) >= ((int32_t)(-2))))) || ((fr_motor_left_cpy) == (fr_motor_right_cpy)));
}

void step(void) {
  (fr_steering_angle_cpy) = (fr_steering_angle);
  (fr_motor_left_cpy) = (fr_motor_left);
  (fr_motor_right_cpy) = (fr_motor_right);
  if ((deadzone_symmetry_violation_0_guard)()) {
    {(deadzone_symmetry_violation)();}
  };
}
